src/rebar_app_discover.erl

```diff
@@ -19,13 +19,14 @@ do(State, LibDirs) ->
     %% Sort apps so we get the same merged deps config everytime
     SortedApps = rebar_utils:sort_deps(Apps),
     lists:foldl(fun(AppInfo, StateAcc) ->
-                        StateAcc1 = merge_deps(AppInfo, StateAcc),
+                        {AppState, StateAcc1} = merge_deps(AppInfo, StateAcc),
                         Name = rebar_app_info:name(AppInfo),
                         OutDir = filename:join(DepsDir, Name),
                         AppInfo1 = rebar_app_info:out_dir(AppInfo, OutDir),
                         ProjectDeps1 = lists:delete(Name, ProjectDeps),
+                        AppInfo2 = rebar_app_info:state(AppInfo1, AppState),
```

Comment:
This step should be done within `merge_deps/2` so I pass in an `AppInfo` and a `State` and get back the same types, rather than having to bake it by hand at every call-site.